Wrappers/Python/cil/optimisation/functions/SAGFunction.py

```diff
@@ -0,0 +1,210 @@
+            self._list_stored_gradients = [
+                0*x for fi in self.functions]
+            self._full_gradient_at_iterate = 0*x
+            self._sampled_grad = x.copy()
+            self._stochastic_grad_difference = x.copy()
+        
+        if self.function_num >= self.num_functions or self.function_num<0 : # check the sampler and raise an error if needed
+            raise IndexError("The sampler has produced an index that exceeds the number of available functions to sample from. Please ensure your sampler only selects from {0,1,...,len(functions)-1} ")
+
+            
+        # Calculate the gradient of the sampled function at the current iterate 
+        self.functions[function_num].gradient(x, out=self._sampled_grad)
+
+        
+        # Calculate the difference between the new gradient of the sampled function and the stored one
+        self._sampled_grad.sapyb(
+            1., self._list_stored_gradients[function_num], -1., out=self._stochastic_grad_difference)
+
+        # Calculate the  approximate gradient
+        out = self._update_approx_gradient(out)
+
+        # Update the stored gradients 
+        self._list_stored_gradients[function_num].fill(
+            self._sampled_grad)
+        
+        # Calculate the stored full gradient
+        self._full_gradient_at_iterate.sapyb(
+            1., self._stochastic_grad_difference, 1., out=self._full_gradient_at_iterate)
+
+        return out
+    
+    def _update_approx_gradient(self, out):
+        """Internal function used to differentiate between the SAG and SAGA calculations. This is the SAG approximation: """
+        if out is None:
+            out = self._stochastic_grad_difference.sapyb(
+                1., self._full_gradient_at_iterate, 1.)
+        else:
+            self._stochastic_grad_difference.sapyb(
+                1., self._full_gradient_at_iterate, 1., out=out)
+
+        return out 
+    
+    def warm_start_approximate_gradients(self, initial):
+        """A function to warm start SAG or SAGA algorithms by initialising all the gradients at an initial point. Equivalently setting :math:`g_i^0=\nabla f_i(x_0)` for initial point :math:`x_0`. 
+        
+        Parameters
+        ----------
+        initial: DataContainer,
+            The initial point to warmstart the calculation
+            
+        Note
+        ----
+        When using SAG or SAGA with a deterministic algorithm, you should warm start the SAG-SAGA Function with the same initial point that you initialise the algorithm
+        
+        """
+        self._list_stored_gradients = [
+            fi.gradient(initial) for fi in self.functions]
+        self._full_gradient_at_iterate = np.sum(self._list_stored_gradients)
+        self._update_data_passes_indices(list(range(self.num_functions)))
+        self._sampled_grad = initial.copy()
+        self._stochastic_grad_difference = initial.copy()
+
+    @property
+    def data_passes_indices(self): 
```

Comment:
Do you want this here or in the base?

Reply:
I think that the initial call to `self._update_data_passes_indices` is different if you warm start or not so it is worth leaving here to avoid another if statement.